rust/connlib/tunnel/src/tests.rs

```diff
@@ -2129,6 +2146,62 @@ fn resolved_ips() -> impl Strategy<Value = HashSet<IpAddr>> {
     collection::hash_set(any::<IpAddr>(), 1..4)
 }
 
+fn non_wildcard_dns_resource() -> impl Strategy<Value = Transition> {
+    (dns_resource(), resolved_ips()).prop_map(|(resource, resolved_ips)| {
+        Transition::AddDnsResource {
+            records: HashMap::from([(resource.address.parse().unwrap(), resolved_ips)]),
+            resource,
+        }
+    })
+}
+
+fn star_wildcard_dns_resource() -> impl Strategy<Value = Transition> {
+    dns_resource().prop_flat_map(|r| {
+        let wildcard_address = format!("*.{}", r.address);
+
+        let records = subdomain_records(r.address, domain_name(1..3));
+        let resource = Just(ResourceDescriptionDns {
+            address: wildcard_address,
+            ..r
+        });
+
+        (resource, records)
+            .prop_map(|(resource, records)| Transition::AddDnsResource { records, resource })
+    })
+}
+
+fn question_mark_wildcard_dns_resource() -> impl Strategy<Value = Transition> {
+    dns_resource().prop_flat_map(|r| {
+        let wildcard_address = format!("?.{}", r.address);
+
+        let records = subdomain_records(r.address, domain_label());
+        let resource = Just(ResourceDescriptionDns {
+            address: wildcard_address,
+            ..r
+        });
+
+        (resource, records)
+            .prop_map(|(resource, records)| Transition::AddDnsResource { records, resource })
+    })
+}
+
+/// A strategy for generating a set of DNS records all nested under the provided base domain.
+fn subdomain_records(
+    base: String,
+    subdomains: impl Strategy<Value = String>,
+) -> impl Strategy<Value = HashMap<DomainName, HashSet<IpAddr>>> {
+    collection::hash_map(subdomains, resolved_ips(), 1..4).prop_map(move |subdomain_ips| {
```

Comment:
something that would be useful to test here is that sometimes the resolved ips for a subdomain will be the same as with another subdomain.

In general, we might want to test, "use some repeated ips" regardless if the domain is the same or not.

Also, for another PR, we should test more than 4 resolved ips as >=5 ips would be a different case from <=4 ips

Reply:
> omething that would be useful to test here is that sometimes the resolved ips for a subdomain will be the same as with another subdomain.
> 
> In general, we might want to test, "use some repeated ips" regardless if the domain is the same or not.

Hmm, this is a bit difficult. I currently exclude any form of overlap. Let me try if we still need that!

Reply:
It appears to work! I removed the precondition that checks for overlaps and it still works so I am guessing it also generates DNS records now that overlap.